lib/RWMutex.ts

```diff
@@ -0,0 +1,216 @@
+   */
+  async rLock() {
+    // provides readable error messages, no need to catch and rethrow
+    const lock = await this._findOrCreateLock(this._lockID);
+    if (lock.readers.indexOf(this._clientID) > -1) {
+      // if this clientID is already a reader, we can re-enter the lock here
+      return;
+    }
+
+    // Loop and do the following:
+    // 1. attempt to acquire the lock (must have no writer)
+    // 2. if not acquired, sleep and retry
+    while (true) {
+      // Acquire the lock
+      try {
+        const result = await this._coll.updateOne({
+          lockID: this._lockID,
+          writer: "",
+        }, {
+          $addToSet: {
+            readers: this._clientID,
+          },
+        });
+        // We check matchedCount rather than modifiedCount here to make the lock re-enterable.
+        // If the lock should not be re-enterable, or clientIDs are not unique, this
+        // implemenation will break.
+        // TODO: option to make it not re-enterable
+        if (result.matchedCount > 0) {
+          return;
+        }
+      } catch (err) {
+        throw new Error(`error aquiring lock: ${err.message}`);
+      }
+
+      await timeoutPromise(this._options.sleepTime);
+    }
+  }
+
+  /*
+   * Unlocks the read lock. Must have the same lock type
+   * @return {Promise} - Resolves when lock is released, rejects if an error occurs
+   */
+  async rUnlock() {
+    let result;
+    try {
+      result = await this._coll.updateOne({
+        lockID: this._lockID,
+        readers: this._clientID,
+      }, {
+        $pull: {
+          readers: this._clientID,
+        },
+      });
+    } catch (err) {
+      throw new Error(`error releasing lock: ${err.message}`);
+    }
+    if (result.matchedCount === 0) {
+      throw new Error(`lock not currently held by client: ${this._clientID}`);
+    }
+    return;
+  }
+
+  /*
+   * Finds or creates the resource in the mongo collection with the specified lockID
+   * @param {string} - unique id of the resource
+   * @return {Promise} - resolves with the lock object, rejects with the formatted error
+   */
+  async _findOrCreateLock(lockID) {
+    let lock;
+    try {
+      lock = await this._coll.find({ lockID: this._lockID }).limit(1).next();
+    } catch (err) {
+      throw new Error(`error finding lock ${lockID}: ${err.message}`);
+    }
+
+    if (!lock) {
+      // lock doesn't exist yet, so we should create it
+      lock = {
+        lockID,
+        readers: [],
+        writer: "",
+      };
+      try {
+        await this._coll.insert(lock);
+      } catch (err) {
+        // Check for E11000 duplicate key error, which means someone inserted the lock before us
+        if (err.code === 11000) {
+          // set the lock to null
+          lock = null;
+        } else {
+          throw new Error(`error creating lock for ${lockID}: ${err.message}`);
+        }
+      }
+    }
+
+    if (!lock) {
+      // handle the duplicate key error, lock must now exist
```

Comment:
I thought about this as well. It _should_ only recurse once, but I have a likely unnecessary fear about an edge case causing infinite recursion. If for some reason (though there's no reason I can think of) find returns null and insert continue to get a duplicate key error, then this function will hang forever and spam the database instead of returning an error.